src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1261,63 +1269,100 @@ private AuthenticationResult localFlow(CallbackHandler callbackHandle,
             return cachedItem;
         }
 
+        // Trying to find refresh token first, if existed, will try to use the refresh token. 
         Logger.v(TAG, "Checking refresh tokens");
         RefreshItem refreshItem = getRefreshToken(request);
+        AuthenticationResult authResult = null;
         if (!promptUser(request.getPrompt()) && refreshItem != null
-                && !StringExtensions.IsNullOrBlank(refreshItem.mRefreshToken)) {
+                && !StringExtensions.IsNullOrBlank(refreshItem.mRefreshToken)) 
+        {
             Logger.v(TAG, "Refresh token is available and it will attempt to refresh token");
-            return refreshToken(callbackHandle, activity, useDialog, request, refreshItem, true);
-        } else {
-            Logger.v(TAG, "Refresh token is not available");
+            try
+            {
+                authResult = getTokenWithRefreshToken(activity, useDialog, request, refreshItem, true);
+            }
+            catch (AuthenticationException authenticationException)
+            {
+                callbackHandle.onError(authenticationException);
+                return null;
+            }
+            
+            if (authResult != null && !StringExtensions.IsNullOrBlank(authResult.getAccessToken()))
+            {
+                callbackHandle.onSuccess(authResult);
+                return authResult;
+            }
```

Comment:
if refresh token request success, and we get a token, then return the token back. The following piece of code will future check if refresh token doesn't exist or refresh token fails to give you the refresh token, if it's non-silent request, we will try with interactive flow.